File: stereoSemiGlobalBlockMatching.py
# Stereo Matching using Semi-Global Block Matching (SGBM)
# Computes a disparity map from a rectified stereo pair using Semi-Global Block Matching

import numpy as np
import cv2 as cv
import matplotlib.pyplot as plt

# Parameters
dispLevels = 16 #disparity range: 0 to dispLevels-1
windowSize = 3
p1 = 5 #occlusion penalty 1
p2 = 10 #occlusion penalty 2

# Load left and right images in grayscale
leftImg = cv.imread("left.png",cv.IMREAD_GRAYSCALE)
rightImg = cv.imread("right.png",cv.IMREAD_GRAYSCALE)

# Apply a Gaussian filter
leftImg = cv.GaussianBlur(leftImg,(5,5),0.6)
rightImg = cv.GaussianBlur(rightImg,(5,5),0.6)

# Get the size
(rows,cols) = leftImg.shape

# Compute pixel-based matching cost
rightImgShifted = np.zeros((rows,cols,dispLevels),dtype=np.int32)
for d in range(dispLevels):
    rightImgShifted[:,d:,d] = rightImg[:,:cols-d]
dataCost = np.absolute(leftImg[:,:,np.newaxis]-rightImgShifted)

# Aggregate the matching cost
dataCost = cv.boxFilter(dataCost,-1,(windowSize,windowSize),normalize=False)

# Compute smoothness cost
d = np.arange(dispLevels)
diff = np.absolute(d-d[np.newaxis,:].T)
p1 = p1*windowSize**2 #normalize p1
p2 = p2*windowSize**2 #normalize p2
smoothnessCost = (diff==1)*p1+(diff>=2)*p2
smoothnessCost3d = smoothnessCost[np.newaxis,:,:].astype(np.int32)

# Initialize path tables for the 8 directions
L1 = np.zeros((rows,cols,dispLevels),dtype=np.int32)
L2 = np.zeros((rows,cols,dispLevels),dtype=np.int32)
L3 = np.zeros((cols,rows,dispLevels),dtype=np.int32)
L4 = np.zeros((cols,rows,dispLevels),dtype=np.int32)
L5 = np.zeros((rows,cols,dispLevels),dtype=np.int32)
L6 = np.zeros((rows,cols,dispLevels),dtype=np.int32)
L7 = np.zeros((rows,cols,dispLevels),dtype=np.int32)
L8 = np.zeros((rows,cols,dispLevels),dtype=np.int32)

# Compute paths for left to right direction
for x in range(1,cols):
    cost = dataCost[:,x-1,:]+L1[:,x-1,:]
    cost = np.amin(cost[:,np.newaxis,:]+smoothnessCost3d,axis=2)
    L1[:,x,:] = cost-np.amin(cost,axis=1)[:,np.newaxis]

# Compute paths for right to left direction
for x in range(cols-2,-1,-1):
    cost = dataCost[:,x+1,:]+L2[:,x+1,:]
    cost = np.amin(cost[:,np.newaxis,:]+smoothnessCost3d,axis=2)
    L2[:,x,:] = cost-np.amin(cost,axis=1)[:,np.newaxis]

# Rotate dataCost for vertical directions
dataCostRotated = np.moveaxis(dataCost,0,1)

# Compute paths for up to down direction
for x in range(1,rows):
    cost = dataCostRotated[:,x-1,:]+L3[:,x-1,:]
    cost = np.amin(cost[:,np.newaxis,:]+smoothnessCost3d,axis=2)
    L3[:,x,:] = cost-np.amin(cost,axis=1)[:,np.newaxis]
L3 = np.moveaxis(L3,0,1)

# Compute paths for down to up direction
for x in range(rows-2,-1,-1):
    cost = dataCostRotated[:,x+1,:]+L4[:,x+1,:]
    cost = np.amin(cost[:,np.newaxis,:]+smoothnessCost3d,axis=2)
    L4[:,x,:] = cost-np.amin(cost,axis=1)[:,np.newaxis]
L4 = np.moveaxis(L4,0,1)

# Edit dataCost for diagonal directions
dataCostEdited1 = np.zeros((rows+cols-1,cols,dispLevels),dtype=np.int32)
dataCostEdited2 = np.zeros((rows+cols-1,cols,dispLevels),dtype=np.int32)
for i in range(cols):
    dataCostEdited1[cols-i-1:rows+cols-i-1,i,:] = dataCost[:,i,:]
    dataCostEdited2[i:rows+i,i,:] = dataCost[:,i,:]

# Initialize temporary tables for diagonal directions
L5a = np.zeros((rows+cols-1,cols,dispLevels),dtype=np.int32)
L6a = np.zeros((rows+cols-1,cols,dispLevels),dtype=np.int32)
L7a = np.zeros((rows+cols-1,cols,dispLevels),dtype=np.int32)
L8a = np.zeros((rows+cols-1,cols,dispLevels),dtype=np.int32)

# Compute paths for left/up to right/down direction
for x in range(1,cols):
    cost = dataCostEdited1[:,x-1,:]+L5a[:,x-1,:]
    cost = np.amin(cost[:,np.newaxis,:]+smoothnessCost3d,axis=2)
    L5a[:,x,:] = cost-np.amin(cost,axis=1)[:,np.newaxis]

# Compute paths for right/down to left/up direction
for x in range(cols-2,-1,-1):
    cost = dataCostEdited1[:,x+1,:]+L6a[:,x+1,:]
    cost = np.amin(cost[:,np.newaxis,:]+smoothnessCost3d,axis=2)
    L6a[:,x,:] = cost-np.amin(cost,axis=1)[:,np.newaxis]

# Compute paths for left/down to right/up direction
for x in range(1,cols):
    cost = dataCostEdited2[:,x-1,:]+L7a[:,x-1,:]
    cost = np.amin(cost[:,np.newaxis,:]+smoothnessCost3d,axis=2)
    L7a[:,x,:] = cost-np.amin(cost,axis=1)[:,np.newaxis]

# Compute paths for right/up to left/down direction
for x in range(cols-2,-1,-1):
    cost = dataCostEdited2[:,x+1,:]+L8a[:,x+1,:]
    cost = np.amin(cost[:,np.newaxis,:]+smoothnessCost3d,axis=2)
    L8a[:,x,:] = cost-np.amin(cost,axis=1)[:,np.newaxis]

# Fill path tables using temporary tables
for i in range(cols):
    L5[:,i,:] = L5a[cols-i-1:rows+cols-i-1,i,:]
    L6[:,i,:] = L6a[cols-i-1:rows+cols-i-1,i,:]
    L7[:,i,:] = L7a[i:rows+i,i,:]
    L8[:,i,:] = L8a[i:rows+i,i,:]

# Compute total cost
S = L1 + L2 + L3 + L4 + L5 + L6 + L7 + L8

# Compute the disparity map
dispMap = np.argmin(S,axis=2)

# Normalize the disparity map for display
scaleFactor = 256/dispLevels
dispImg = (dispMap*scaleFactor).astype(np.uint8)

# Show disparity map
plt.imshow(dispImg,cmap="gray")
plt.show(block=False)
plt.pause(0.01)

# Save disparity map
cv.imwrite("disparity.png",dispImg)

plt.show()

File: stereoSemiGlobalBlockMatching.m
% Stereo Matching using Semi-Global Block Matching (SGBM)
% Computes a disparity map from a rectified stereo pair using Semi-Global Block Matching

% Set parameters
dispLevels = 16; %disparity range: 0 to dispLevels-1
windowSize = 3;
p1 = 5; %occlusion penalty 1
p2 = 10; %occlusion penalty 2

% Define data cost computation
dataCostComputation = @(differences) abs(differences); %absolute differences
%dataCostComputation = @(differences) differences.^2; %square differences

% Define smoothness cost computation
smoothnessCostComputation = @(differences) (abs(differences)==1)*p1+(abs(differences)>=2)*p2;

% Load left and right images in grayscale
leftImg = rgb2gray(imread('left.png'));
rightImg = rgb2gray(imread('right.png'));

% Apply a Gaussian filter
leftImg = imgaussfilt(leftImg,0.6,'FilterSize',5);
rightImg = imgaussfilt(rightImg,0.6,'FilterSize',5);

% Get the size
[rows,cols] = size(leftImg);

% Compute pixel-based matching cost (data cost)
rightImgShifted = zeros(rows,cols,dispLevels,'int32');
for d = 0:dispLevels-1
    rightImgShifted(:,d+1:end,d+1) = rightImg(:,1:end-d);
end
dataCost = dataCostComputation(int32(leftImg)-rightImgShifted);

% Aggregate the matching cost
dataCost = int32(convn(dataCost,ones(windowSize,windowSize,1),'same'));

% Compute smoothness cost
d = 0:dispLevels-1;
smoothnessCost = smoothnessCostComputation(d-d.')*windowSize^2; %normalize
smoothnessCost3d = zeros(1,dispLevels,dispLevels,'int32');
smoothnessCost3d(1,:,:) = smoothnessCost;

% Initialize path tables for the 8 directions
L1 = zeros(rows,cols,dispLevels,'int32');
L2 = zeros(rows,cols,dispLevels,'int32');
L3 = zeros(cols,rows,dispLevels,'int32');
L4 = zeros(cols,rows,dispLevels,'int32');
L5 = zeros(rows,cols,dispLevels,'int32');
L6 = zeros(rows,cols,dispLevels,'int32');
L7 = zeros(rows,cols,dispLevels,'int32');
L8 = zeros(rows,cols,dispLevels,'int32');

% Compute paths for left to right direction
for x = 2:cols
    cost = dataCost(:,x-1,:)+L1(:,x-1,:);
    cost = min(cost+smoothnessCost3d,[],3);
    L1(:,x,:) = cost-min(cost,[],2);
end

% Compute paths for right to left direction
for x = cols-1:-1:1
    cost = dataCost(:,x+1,:)+L2(:,x+1,:);
    cost = min(cost+smoothnessCost3d,[],3);
    L2(:,x,:) = cost-min(cost,[],2);
end

% Rotate dataCost for vertical directions
dataCostRotated = permute(dataCost,[2 1 3]);

% Compute paths for up to down direction
for x = 2:rows
    cost = dataCostRotated(:,x-1,:)+L3(:,x-1,:);
    cost = min(cost+smoothnessCost3d,[],3);
    L3(:,x,:) = cost-min(cost,[],2);
end
L3 = permute(L3,[2 1 3]);

% Compute paths for down to up direction
for x = rows-1:-1:1
    cost = dataCostRotated(:,x+1,:)+L4(:,x+1,:);
    cost = min(cost+smoothnessCost3d,[],3);
    L4(:,x,:) = cost-min(cost,[],2);
end
L4 = permute(L4,[2 1 3]);

% Edit dataCost for diagonal directions
dataCostEdited1 = zeros(rows+cols-1,cols,dispLevels,'int32');
dataCostEdited2 = zeros(rows+cols-1,cols,dispLevels,'int32');
for i = 1:cols
    dataCostEdited1(cols-i+1:rows+cols-i,i,:) = dataCost(:,i,:);
    dataCostEdited2(i:rows+i-1,i,:) = dataCost(:,i,:);
end

% Initialize temporary tables for diagonal directions
L5a = zeros(rows+cols-1,cols,dispLevels,'int32');
L6a = zeros(rows+cols-1,cols,dispLevels,'int32');
L7a = zeros(rows+cols-1,cols,dispLevels,'int32');
L8a = zeros(rows+cols-1,cols,dispLevels,'int32');

% Compute paths for left/up to right/down direction
for x = 2:cols
    cost = dataCostEdited1(:,x-1,:)+L5a(:,x-1,:);
    cost = min(cost+smoothnessCost3d,[],3);
    L5a(:,x,:) = cost-min(cost,[],2);
end

% Compute paths for right/down to left/up direction
for x = cols-1:-1:1
    cost = dataCostEdited1(:,x+1,:)+L6a(:,x+1,:);
    cost = min(cost+smoothnessCost3d,[],3);
    L6a(:,x,:) = cost-min(cost,[],2);
end

% Compute paths for left/down to right/up direction
for x = 2:cols
    cost = dataCostEdited2(:,x-1,:)+L7a(:,x-1,:);
    cost = min(cost+smoothnessCost3d,[],3);
    L7a(:,x,:) = cost-min(cost,[],2);
end

% Compute paths for right/up to left/down direction
for x = cols-1:-1:1
    cost = dataCostEdited2(:,x+1,:)+L8a(:,x+1,:);
    cost = min(cost+smoothnessCost3d,[],3);
    L8a(:,x,:) = cost-min(cost,[],2);
end

% Fill path tables using temporary tables
for i = 1:cols
    L5(:,i,:) = L5a(cols-i+1:rows+cols-i,i,:);
    L6(:,i,:) = L6a(cols-i+1:rows+cols-i,i,:);
    L7(:,i,:) = L7a(i:rows+i-1,i,:);
    L8(:,i,:) = L8a(i:rows+i-1,i,:);
end

% Compute total cost
S = L1 + L2 + L3 + L4 + L5 + L6 + L7 + L8;

% Compute the disparity map
[~,ind] = min(S,[],3);
dispMap = ind-1;

% Normalize the disparity map for display
scaleFactor = 256/dispLevels;
dispImg = uint8(dispMap*scaleFactor);

% Show disparity map
figure; imshow(dispImg)

% Save disparity map
imwrite(dispImg,'disparity.png')